pkg/receive/handler.go

```diff
@@ -0,0 +1,222 @@
+}
+
+func instrumentHandler(handlerName string, handler http.HandlerFunc) http.HandlerFunc {
+	return promhttp.InstrumentHandlerDuration(
+		requestDuration.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+		promhttp.InstrumentHandlerResponseSize(
+			responseSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+			handler,
+		),
+	)
+}
+
+func NewHandler(logger log.Logger, o *Options) *Handler {
+	router := route.New().WithInstrumentation(instrumentHandler)
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	h := &Handler{
+		logger:       logger,
+		router:       router,
+		readyStorage: o.ReadyStorage,
+		context:      o.Context,
+		receiver:     o.Receiver,
+		options:      o,
+		quitCh:       make(chan struct{}),
+	}
+
+	readyf := h.testReady
+	router.Post("/receive", readyf(h.receive))
+
+	return h
+}
+
+// Ready sets Handler to be ready.
+func (h *Handler) Ready() {
+	atomic.StoreUint32(&h.ready, 1)
+}
+
+// Verifies whether the server is ready or not.
+func (h *Handler) isReady() bool {
+	ready := atomic.LoadUint32(&h.ready)
+	return ready > 0
+}
+
+// Checks if server is ready, calls f if it is, returns 503 if it is not.
+func (h *Handler) testReady(f http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if h.isReady() {
+			f(w, r)
+		} else {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			_, err := fmt.Fprintf(w, "Service Unavailable")
+			if err != nil {
+				h.logger.Log("msg", "failed to write to response body", "err", err)
+			}
+		}
+	}
+}
+
+// Quit returns the receive-only quit channel.
+func (h *Handler) Quit() <-chan struct{} {
+	return h.quitCh
+}
+
+// Checks if server is ready, calls f if it is, returns 503 if it is not.
+func (h *Handler) testReadyHandler(f http.Handler) http.HandlerFunc {
+	return h.testReady(f.ServeHTTP)
+}
+
+// Run serves the HTTP endpoints.
+func (h *Handler) Run(ctx context.Context) error {
+	level.Info(h.logger).Log("msg", "Start listening for connections", "address", h.options.ListenAddress)
+
+	listener, err := net.Listen("tcp", h.options.ListenAddress)
+	if err != nil {
+		return err
+	}
+	listener = netutil.LimitListener(listener, h.options.MaxConnections)
+
+	// Monitor incoming connections with conntrack.
+	listener = conntrack.NewListener(listener,
+		conntrack.TrackWithName("http"),
+		conntrack.TrackWithTracing())
+
+	var (
+		m     = cmux.New(listener)
+		httpl = m.Match(cmux.HTTP1Fast())
+	)
+
+	operationName := nethttp.OperationNameFunc(func(r *http.Request) string {
+		return fmt.Sprintf("%s %s", r.Method, r.URL.Path)
+	})
+	mux := http.NewServeMux()
+	mux.Handle("/", h.router)
+
+	errlog := stdlog.New(log.NewStdlibAdapter(level.Error(h.logger)), "", 0)
+
+	httpSrv := &http.Server{
+		Handler:     nethttp.Middleware(opentracing.GlobalTracer(), mux, operationName),
+		ErrorLog:    errlog,
+		ReadTimeout: h.options.ReadTimeout,
+	}
+
+	errCh := make(chan error)
```

Comment:
I am pretty sure we leak go routine here, as we send twice but read only once.

Worth to use `run.Group` here for easier code?

Reply:
Yes, this should be a `run.Group`.

Reply:
still valid?

Reply:
Not addressed